### PR TITLE
Implement if-let match guards

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
Implements rust-lang/rfcs#2294 (tracking issue: #51114).

I probably should do a few more things before this can be merged:
- [x] Add tests (added basic tests, more advanced tests could be done in the future?)
- [x] Add lint for exhaustive if-let guard (comparable to normal if-let statements)
- [x] Fix clippy

However since this is a nightly feature maybe it's fine to land this and do those steps in follow-up PRs.

Thanks a lot @matthewjasper :heart: for helping me with lowering to MIR! Would you be interested in reviewing this?
r? @ghost for now

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 1]
```diff
@@ -1723,15 +1727,41 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+                    let scrutinee_span = scrutinee.span();
+                    let scrutinee_place = unpack!(block = self.lower_scrutinee(block, scrutinee.clone(), scrutinee_span));
+                    let mut guard_candidate = Candidate::new(scrutinee_place, &pat, false);
+                    let wildcard = Pat::wildcard_from_ty(pat.ty);
+                    let mut otherwise_candidate = Candidate::new(scrutinee_place, &wildcard, false);
+                    let fake_borrow_temps =
+                        self.lower_match_tree(block, pat.span, false, &mut [&mut guard_candidate, &mut otherwise_candidate]);
+                    self.declare_bindings(
+                        None,
+                        pat.span,
```
Comment (LeSeulArtichaut) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:1746]: Note to self: you set this just to see if it works, don't forget to change it!

**Comment by LeSeulArtichaut** [comment]
@matthewjasper This should be ready for review now!

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 2]
```diff
@@ -355,6 +365,41 @@ fn irrefutable_let_pattern(tcx: TyCtxt<'_>, span: Span, id: HirId, source: hir::
...
+            cx.tcx,
+            pat.span,
+            pat_id,
+            hir::MatchSource::IfLetDesugar { contains_else_clause: true },
+        ),
+        Useful(_) => {}
+        UsefulWithWitness(_) => bug!(),
+    }
+}
+
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:402]: Soryy, I did a bit of a refactor in https://github.com/rust-lang/rust/pull/78995. I think you need something like this now:

```rust
fn check_if_let_guard<'p, 'tcx>(
    cx: &mut MatchCheckCtxt<'p, 'tcx>,
    pat: &'p super::Pat<'tcx>,
    pat_id: HirId,
) {
    let arms = [MatchArm { pat, hir_id: pat_id, has_guard: false }];
    let report = compute_match_usefulness(&cx, &arms, pat_id, pat.ty);

    match &report.arm_usefulness[0] {
        NotUseful => unreachable_pattern(cx.tcx, pat.span, pat_id, None),
		Useful(unreachables) if unreachables.is_empty() => {}
        Useful(unreachables) => {
            let mut unreachables: Vec<_> = unreachables.into_iter().flatten().collect();
            // Emit lints in the order in which they occur in the file.
            unreachables.sort_unstable();
            for span in unreachables {
                unreachable_pattern(cx.tcx, span, pat_id, None);
            }
        }
        UsefulWithWitness(_) => bug!(),
    }

    if report.non_exhaustiveness_witnesses.is_empty() {
        // The match is exhaustive, i.e. the if let pattern is irrefutable.
        irrefutable_let_pattern(
            cx.tcx,
            pat.span,
            pat_id,
            hir::MatchSource::IfLetDesugar { contains_else_clause: true },
        )
    }
}
```

The printing of unreachable subpatterns is duplicate with what's happening in `report_arm_reachability`. Is there a way you could deduplicate it? Maybe adding a `MatchSource::IfLetGuardDesugar`?

**Comment by LeSeulArtichaut** [comment]
Thanks a lot @Nadrieril for the pointers!
@matthewjasper This is ready for review again, sorry for the time I took to rebase this PR.

File: src/tools/clippy/clippy_lints/src/utils/author.rs [hunk 3]
```diff
@@ -372,6 +372,9 @@ impl<'tcx> Visitor<'tcx> for PrintVisitor {
                                 self.current = if_expr_pat;
                                 self.visit_expr(if_expr);
                             },
+                            hir::Guard::IfLet(..) => {
+                                unimplemented!("#[clippy::author] not yet implemented for if-let guards")
```
Comment (matthewjasper) [inline src/tools/clippy/clippy_lints/src/utils/author.rs:376]: This should probably be implemented, something like this:
```rust
hir::Guard::IfLet(ref if_let_pat, ref if_let_expr) => {
    let if_let_pat_pat = self.next("pat");
    let if_let_expr_pat = self.next("expr");
    println!(
        "    if let Guard::IfLet(ref {}, ref {}) = {};",
        if_let_pat_pat, if_let_expr_pat, guard_pat
    );
    self.current = if_let_expr_pat;
    self.visit_expr(if_let_expr);
    self.current = if_let_pat_pat;
    self.visit_pat(if_let_pat);
}
```

**Comment by LeSeulArtichaut** [comment]
All right, CI passes again now 😅

**Comment by LeSeulArtichaut** [comment]
Failure looks spurious?

File: src/test/ui/rfc-2294-if-let-guard/bindings.rs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#![feature(if_let_guard)]
+#![allow(incomplete_features)]
+
+fn main() {
+    match Some(None) {
+        Some(x) if let Some(y) = x => (x, y),
+        _ => y, //~ ERROR cannot find value `y`
+    }
+    y //~ ERROR cannot find value `y`
```
Comment (camelid) [inline src/test/ui/rfc-2294-if-let-guard/bindings.rs:9]: Shouldn't this define an `x` and a `y` so this test can be `check-pass`? That way we can be absolutely certain that it works.

Comment (camelid) [reply src/test/ui/rfc-2294-if-let-guard/bindings.rs:9]: Oops, I just realized it's intentional that `y`'s not defined ;)

File: src/test/ui/rfc-2294-if-let-guard/feature-gate.stderr [hunk 5]
```diff
@@ -17,7 +17,7 @@ LL |         () if let 0 = 1 => {}
    = help: add `#![feature(if_let_guard)]` to the crate attributes to enable
 
 error[E0658]: `if let` guard is not implemented
```
Comment (camelid) [inline src/test/ui/rfc-2294-if-let-guard/feature-gate.stderr:19]: This should probably say "`if let` guard is experimental".

**Comment by jyn514** [comment]
Looks like aarch64-gnu was cancelled on Github's end? There's no logs.

@bors retry

File: src/test/ui/rfc-2294-if-let-guard/feature-gate.stderr (hunk 5 above)
Comment (LeSeulArtichaut) [reply src/test/ui/rfc-2294-if-let-guard/feature-gate.stderr:19]: I'll file a PR fixing this later today

**Comment by Mark-Simulacrum** [comment]
This was a regression of 3% on the match-stress-enum benchmark, but since it's directly adding more code to it that seems relatively expected. https://perf.rust-lang.org/compare.html?start=a6491be5be9344a325b7e49b0114f3cf67ef199e&end=9b84d36a0b9ea3bf305f36f08d50aa42c26f96c2&stat=instructions:u

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 6]
```diff
@@ -164,10 +164,20 @@ impl<'tcx> MatchVisitor<'_, 'tcx> {
...
         }
 
         let mut cx = self.new_cx(scrut.hir_id);
 
+        for arm in arms {
+            if let Some(hir::Guard::IfLet(ref pat, _)) = arm.guard {
+                let tpat = self.lower_pattern(&mut cx, pat, &mut false).0;
+                check_if_let_guard(&mut cx, &tpat, pat.hir_id);
+            }
+        }
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:179]: Looks like this loop could be merged with the one above, that would probably reduce the perf impact

Comment (camelid) [reply compiler/rustc_mir_build/src/thir/pattern/check_match.rs:179]: I'll try opening a PR.

**Pull request merged**